### PR TITLE
Migrate build system to Gradle

### PR DESCRIPTION
**Pull request by vlsi**
Sample commands:

```
./gradlew # shows help
./gradlew tasks
./gradlew build
./gradlew test
./gradlew jar
```

I use https://github.com/dougborg/gdub so I can execute `gw jar` from any directory (e.g. from `core` directory)

See https://github.com/javacc/javacc/pull/143

Sample GitHub Actions: https://github.com/vlsi/javacc21/actions/runs/43225096
Sample deployment to Central: https://oss.sonatype.org/content/repositories/comjavacc-1000/com/javacc/javacc-core/21.3.0/

Here's how releasing to Central looks like:

```
$ gw prepareVote -Prc=1 -Pgh

...
...

> Task :javacc-core:javadoc
javacc21/core/build/javacc/javacc/javacc/parser/JavaCharStream.java:166: warning - @Deprecated is an unknown tag -- same as a known tag except for case.
javacc21/core/build/javacc/javacc/javacc/parser/JavaCharStream.java:174: warning - @Deprecated is an unknown tag -- same as a known tag except for case.
2 warnings

> Task :javacc-core:initializeNexusStagingRepository
Initialized stagingRepositoryId comjavacc-1000 for repository nexus

> Task :closeRepository
Requested operation wasn't successful in first try. Retrying maximum 20 times with 2 seconds delay between attempts.
Requested operation was executed successfully in attempt 16 (maximum allowed 21)

> Task :pushRcTag
Pushing tag to Git remote release-origin: https://github.com/vlsi/javacc.git
Message from release-origin:
  refs/tags/v21.3.0-rc1: OK, 7df71b13342bfab9c50a9546400e8ed6e9558f75 (fastForward)

> Task :generateVoteText
Please find draft vote text in javacc21/build/prepareVote/mail.txt

> Task :prepareVote
javacc v21.3.0-rc1 is ready for preview.

Git SHA: 01b5a5270fcf75fca6883a83e22f61327f555aae
Staging repository: https://oss.sonatype.org/content/repositories/comjavacc-1000

BUILD SUCCESSFUL in 3m 33s
16 actionable tasks: 16 executed
```

### CODE REVIEW

**Comment by revusky** [comment]
Well, generally speaking, I am delighted that you put in the work to do this. I tend to think moving over to Gradle is a good idea, but the truth is that, on my own steam, it would never happen. Not because I think it is a bad idea or anything, but just because I would always have more stimulating things to do. 

So, probably, tomorrow I will just merge all this in. Just a couple of comments though.

First of all, do we really have to call the directory with the templates in it "resources"? Truth told, I just hate these weird ways of using words. To me, a "resource" would be crude oil under the ground or something like that. I really think that whoever started this thing of calling files "resources" deserves to be shot. Along with the person who came up with the idea of calling the files you upload to the Maven repository "artifacts". To me, an artifact would be something like an old utensil you find in an archaeological excavation. (And then, for some reason on Github, it is called an "asset", like real estate or shares in a company or something.) So, I would just prefer to call templates templates, not "resources". 

That said, it is not very important to me either probably, and if these are just the standard naming conventions, well, okay. I guess it's just not worth fighting against it. It's like trying to insist that the word "gay" just means happy or cheerful. The battle is already lost. The other thing is that there is this tendency for things to get buried very deeply in sub-sub-directories, like src/main/java/javacc and so on, which I honestly find annoying. But I was thinking that, finally, since things are already buried so deep, maybe we should just move all the javacc.* package naming to com.javacc.* since I do own the javacc.com directory now. Of course, that leads to files being buried one layer deeper,  like src/main/java/com/javacc/parser and so on, but since it is already so many layers deep, what's the difference now? 

As a straight practical question, it does not make much difference, but I was just thinking about maybe biting the bullet now and moving to a com.javacc.* naming. I really don't know. It can be  annoying to have to dig down into so many subdirectories to get at the file you want to look at, but I mention it because if we're going for the most standard directory layout, then maybe... Though, I suppose there is no need to decide that now anyway. Moving packages around is just standard refactoring that Eclipse or IntelliJ can do for you automtically, isn't it? But I guess if we were going to do that, might as well do it now.

But anyway, thanks again for the contribution.

File: README.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+[![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.javacc/javacc-core/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.javacc/javacc-core)
```
Comment (angelozerr) [inline README.md]: Are you sure it's the proper URL? It seems the URL is for javacc not for javacc21?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
...
+[![Travis Build Status](https://travis-ci.org/javacc21/javacc21.svg?branch=master)](https://travis-ci.org/javacc21/javacc21)
+[![CI Status](https://github.com/javacc21/javacc21/workflows/CI/badge.svg)](https://github.com/javacc21/javacc21/actions)
+[![Javadocs](https://www.javadoc.io/badge/com.javacc/javacc-core.svg)](https://www.javadoc.io/doc/com.javacc/javacc-core)
```
Comment (angelozerr) [inline README.md]: Are you sure it's the proper URL? It seems the URL is for javacc not for javacc21?

File: README.md (hunk 1 above)
Comment (vlsi) [reply README.md]: javacc21 is com.javacc

File: README.md (hunk 2 above)
Comment (vlsi) [reply README.md]: javacc21 is com.javacc, so the URL is right.

Comment (angelozerr) [reply README.md]: Ok thanks for the info.

File: README.md (hunk 1 above)
Comment (angelozerr) [reply README.md]: Ok thanks for the info.

**Comment by vlsi** [comment]
I've renamed the package to `com.javacc.`

It is not clear what to do with `freemarker.` though.

**Comment by revusky** [comment]
Okay, I was sort of wavering on the question of changing the `javacc.*` package name to `com.javacc.*` but finally I was tending to think we might as well just do it. (I was probably about +0.1 on it.)  I infer  that, in your local version, all the packages are moved and all the existing build/test is working. So, that's fine. That's resolved then. Great!

As for `freemarker.*`, I think it would be a quite bad idea to change the package names. FreeMarker has a long history and is a library being used in many projects. So, breaking all kinds of existing code that is sitting there running that has `import freemarker.template.*;` up top, to do that for basically zero practical benefit.... Actually, one notes that when FreeMarker (the obsolete 2.3 branch.. I told you the history...) became an Apache project, they did not change the package names from `freemarker.*` to `org.apache.freemarker.*`. Though they did mavenize the whole directory structure, by the way, so what was previously located at `src/freemarker/template` is now in `src/main/java/freemarker/template` (*Of course they did! Because we all know how wonderfully beneficial that is!* (Excuse the sarcasm..)) But, anyway, they didn't change the package names. That just would not have been a practical thing to do. With this JavaCC project, OTOH, there is quite a large degree of freedom right now to change packages or rename things generally really. It's a completely different situation.

Well, at some point, if you are motivated, you could see about mavenizing/gradlizing [the project here](https://github.com/freemarker/freemarker3). I suppose once we have our "artifacts" in the Maven repository, we can just grab the latest freemarker.jar and javacc.jar from Maven central to bootstrap the project.

Anyway, once you have this in what you feel is a final state (maybe you do already) I'll just merge the pull request. Basically, I'll just assume that you know what you're doing and not second-guess it. (And besides, in the worst case imaginable, I did tag a release yesterday so I could easily revert to that, but I don't anticipate that happening...)

**Comment by angelozerr** [comment]
>  First of all, do we really have to call the directory with the templates
in it "resources"?

@Jonathan Revusky <revusky@gmail.com>  it's a standard layout used in maven
and gradle. I mean:

 * /src
      /main
         /java
         /resources
      /test
         /java
         /resources

IMHO, you should follow this standard layout.

File: core/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    implementation(files("$rootDir/bin/freemarker.jar"))
+}
+
+val javacc by tasks.registering(JavaCCTask::class) {
+    description = "Generate the Java CC Main Parser"
+    // For now the grammar has to be co-located with template files, so they get preference
+    // over the files inside javacc.jar
+    inputFile.set(file("src/main/resources/javacc/output/java/JavaCC.javacc"))
+}
+
```
[the remark this answers is not in the document]
Comment (vlsi) [reply core/build.gradle.kts]: I assume this is answered in the mail.

>  you just built, then you just copy the javacc.jar you just built to the bin directory and build again. And then you are using the newer templates. You see my point here?

I do not see the point in manually copying files all over the place :)

There is a previous version (e.g. `bin/javacc.jar`). The build produces the next jar. That is it.

> seeing whether I can rebuild. (And retest!)

Re-build should probably produce the same byte-by-byte results. I have not tried that, but it should be the case.

**Comment by vlsi** [comment]
Note: I have not migrated tests/examples, and I guess they can be updated later.

The key point, for now, is loading the project to IDE, and releasing versions to Central.

Examples should probably be located in a different repository: it keeps the samples small, however, it requires javacc to be downloadable from Central.

Here's how samples might be organized: https://github.com/junit-team/junit5-samples/

**Pull request merged**